server/memphis_handler_schemas.go

```diff
@@ -0,0 +1,368 @@
+
+	"github.com/gin-gonic/gin"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+type SchemasHandler struct{ S *Server }
+type SchemasVersionsHandler struct{ S *Server }
+
+const (
+	schemaObjectName = "Schema"
+)
+
+var (
+	validKeyRe              = regexp.MustCompile(`\A[-/_=\.a-zA-Z0-9]+\z`)
+	ErrObjectConfigRequired = errors.New("nats: object-store config required")
+	ErrBadObjectMeta        = errors.New("nats: object-store meta information invalid")
+	ErrObjectNotFound       = errors.New("nats: object not found")
+	ErrInvalidStoreName     = errors.New("nats: invalid object-store name")
+	ErrInvalidObjectName    = errors.New("nats: invalid object name")
+	ErrDigestMismatch       = errors.New("nats: received a corrupt object, digests do not match")
+	ErrNoObjectsFound       = errors.New("nats: no objects found")
+)
+
+func validateSchemaName(schemaName string) error {
+	return validateName(schemaName, schemaObjectName)
+}
+
+func validateSchemaType(schemaType string) error {
+	invalidTypeErrStr := fmt.Sprintf("%v should be protobuf", schemaType)
+	invalidTypeErr := errors.New(invalidTypeErrStr)
+
+	if schemaType == "protobuf" {
+		return nil
+	} else {
+		return invalidTypeErr
+	}
+
+}
+
+func validateSchemaContent(schemaContent string) error {
+	invalidSchemaContentErrStr := fmt.Sprintf("%v should be protobuf", schemaContent)
+	invalidSchemaContentErr := errors.New(invalidSchemaContentErrStr)
+	if strings.Contains(schemaContent, "syntax") {
+		return nil
+
+	} else {
+		return invalidSchemaContentErr
+
+	}
+}
+
+func (sh SchemasHandler) GetSchemaDetailsBySchemaName(schemaName string, versionNumber int) (models.ExtendedSchemaDetails, error) {
+	var schema models.Schema
+	err := schemasCollection.FindOne(context.TODO(), bson.M{
+		"name": schemaName,
+	}).Decode(&schema)
+
+	if err != nil {
+		return models.ExtendedSchemaDetails{}, err
+	}
+	var schemaVersion models.SchemaVersion
+	filter := bson.M{
+		"version_number": versionNumber,
+		"$or": []interface{}{
+			bson.M{"_id": bson.M{"$in": schema.Versions}},
+		},
+	}
+
+	err = schemasVersionCollection.FindOne(context.TODO(), filter).Decode(&schemaVersion)
+	if err != nil {
+		return models.ExtendedSchemaDetails{}, err
+	}
+
+	extedndedSchemaDetails := models.ExtendedSchemaDetails{
+		ID:            schema.ID,
+		SchemaName:    schema.Name,
+		VersionNumber: schemaVersion.VersionNumber,
+		Active:        schemaVersion.Active,
+		CreatedByUser: schemaVersion.CreatedByUser,
+		Type:          schema.Type,
+		SchemaContent: schemaVersion.SchemaContent,
+	}
+	return extedndedSchemaDetails, nil
+}
+
+func (sh SchemasHandler) GetAllSchemasDetails() ([]models.ExtendedSchema, error) {
+	var schemas []models.Schema
+	cursor, err := schemasCollection.Aggregate(context.TODO(), mongo.Pipeline{
+		bson.D{{"$match", bson.D{{"$or", []interface{}{
+			bson.D{{"is_deleted", false}},
```

Comment:
we said that remove schema will remove the document so is_deleted is not relevant